### DISCUSSION TITLE
How do I represent a transient source?

### DISCUSSION THREAD

**Discussion by Sandip433** [usnistgov/fipy#1036] [Q&A] [answered]
As per the Fipy documentation, The time-rate-of change of an independent variable $\xi$, such as in 
$\frac{\partial (\rho_1 \cdot \phi)}{\partial t} = \frac{\partial (\rho_2 \cdot \xi)}{\partial t}$
can be written as 
TransientTerm(coeff=rho1, var=phi) == rho2 * (xi - xi.old) / timeStep
I have some questions
(1) Can I write the above as 
TransientTerm(coeff=rho1, var=phi) == TransientTerm(coeff=rho2, var=xi)
after declaring $\xi$ as cellvariable ? Would that be right ?
Or
Phi_1 = CellVariable(mesh=mesh, value=rho1 * phi - rho2 * xi)
TransientTerm(coeff=1,var=Phi_1)==0
(2) Here I think that $\xi$ is a funtion of time, not a cellvariable. What if $\xi$ is a cellvariable and function of $\phi$ (cellvariable) as well as time ?
Do I have to use chain rule of differentiation ? If yes, how can I do it ?

**Comment by guyer** [comment]
> (1) Can I write the above as TransientTerm(coeff=rho1, var=phi) == TransientTerm(coeff=rho2, var=xi) after declaring ξ as cellvariable ? Would that be right ? 

If $\phi$ and $\xi$ are both functions of space and time, then yes, you can do this. Whether you'll get reasonable solutions will depend on what other equation you are solving (you must have two equations for two unknowns).

> Or Phi_1 = CellVariable(mesh=mesh, value=rho1 * phi - rho2 * xi) 

This will only assign the instantaneous value of `rho1 * phi - rho2 * xi` to `Phi_1`. It won't update if `phi` and `xi` change. Instead, just write
```python
Phi_1 = rho1 * phi - rho2 * xi
```
**However**
> TransientTerm(coeff=1,var=Phi_1)==0

an operator variable like `Phi_1 = rho1 * phi - rho2 * xi` cannot be used as a solution variable in a FiPy equation, so stick with
```python
TransientTerm(coeff=rho1, var=phi) == TransientTerm(coeff=rho2, var=xi)
```

> (2) Here I think that ξ is a funtion of time, not a cellvariable. 

If $\xi$ is not a `CellVariable`, i.e., not a function of space, then you cannot use it in `ransientTerm(coeff=rho2, var=xi)`.  You would need to do the explicit `rho2 * (xi - xi.old) / timeStep`, except only a `CellVariable` has an `.old` property, so you would need to manually maintain a simple pair of `Variable` objects for `xi` and `xi_old`. It would probably be easier to make `xi` a `CellVariable` and simply ensure that its governing equation(s) have no spatial dependencies.

> What if ξ is a cellvariable and function of ϕ (cellvariable) as well as time ? Do I have to use chain rule of differentiation ? If yes, how can I do it ?

Explicit function or implicit?

If you know the functional form of $\xi$, then you should just write
```python
xi = func(phi, t)
```
and then use `xi` wherever you need it. `xi` will not get a governing PDE. If you need the time derivative of $\xi$, then you'll need to run the chain rule.

If the relationship between $\xi$ and $\phi$ is implicit, i.e., they are coupled, but one is not an explicit function of the other, then simply write their PDEs and either couple them or solve them sequentially. If the equations are well-posed, then FiPy will resolve the self-consistent solution to both of them.

To advise any further, I'll need to see what equations we're talking about.

**Comment by guyer** [comment]
> As per the Fipy documentation, The time-rate-of change of an independent variable ξ, such as in ∂(ρ1⋅ϕ)∂t=∂(ρ2⋅ξ)∂t can be written as TransientTerm(coeff=rho1, var=phi) == rho2 * (xi - xi.old) / timeStep

FWIW, all of the guidance on how to [represent a … term that *doesn’t* involve the dependent variable](https://pages.nist.gov/fipy/en/latest/FAQ.html#how-do-i-represent-a-term-that-doesn-t-involve-the-dependent-variable) could really stand a refresh to better reflect the availability of coupled terms.

**Comment by Sandip433** [comment]
I want to simulate Laser irradiation on Semiconductor Si. These are the equations
![1](https://github.com/usnistgov/fipy/assets/132244694/51401f8b-fadd-4cdf-a63b-113a56399a62)
Here I have 3 unknown $N_c, T_e, T_l$ and I am having problem to implement equations 3 and 4 and I am not sure which one is more suitable for Fipy because one parameter $E_g$ is a explicit function of $N_c$ and $T_l$ and I want to know how to perform chain rule of differentiation in Fipy. Also I was wondering if $E_g$ is implicitly connected to $N_c, T_e, T_l$ e.g. $sin(T_e + E_g) + log(T_l-E_g^3)=0$ then how to perform time derivative of $E_g$ ?

**Comment by guyer** [comment]
> I am having problem to implement equations 3 and 4 and I am not sure which one is more suitable for Fipy because one parameter Eg is a explicit function of Nc and Tl and I want to know how to perform chain rule of differentiation in Fipy. 

FiPy offers no facility for doing the chain rule. You'd need to do this by hand.

> Also I was wondering if Eg is implicitly connected to Nc,Te,Tl e.g. sin(Te+Eg)+log(Tl−Eg3)=0 then how to perform time derivative of Eg ?

Again, by hand, e.g.

$$
\begin{align}
\sin(T_e + E_g) + \log(T_l + E_g^3) &= 0
\\
\frac{\partial}{\partial t}\left[\sin(T_e + E_g) + \log(T_l + E_g^3)\right] &= \frac{\partial}{\partial t} 0
\\
\cos(T_e + E_g)\left(\frac{\partial T_e}{\partial t} + \frac{\partial E_g}{\partial t}\right)
+\frac{1}{T_l + E_g^3}\left(\frac{\partial T_l}{\partial t} + 3 E_g^2 \frac{\partial E_g}{\partial t}\right) &= 0
\end{align}
$$

**Comment by Sandip433** [comment]
![1](https://github.com/usnistgov/fipy/assets/132244694/438ccd9b-2363-4dc0-8846-be1f7c755bdc)
So to implement equation (4) can I write 
`TransientTerm(coeff=C_e,var=T_e)  =  DiffusionTerm(coeff=k_e,var=T_e) - ImplicitSourceTerm(coeff=3*N_c*k_B/tao_c, var=T_e) +         ImplicitSourceTerm(coeff=3*N_c*k_B/tao_c, var=T_l) + alpha * I + beta * I**2 - E_g * TransientTerm(coeff=1,var=N_c) - N_c * (E_g - E_g.old) / timeStep`
I assume that Fipy will take care of dependence of $T_l$ and $N_c$ in $E_g$ implicitly ?

**Reply by guyer** [reply, marked as answer]
That should work
